### PR TITLE
replace all ctx.fail calls with custom func

### LINKED ISSUE #296: Suppress "Usage:" in stderr when yubikey not found
- YubiKey Manager (ykman) version: 3.1.0
- How was it installed?: Homebrew
- Operating system and version: MacOS Catalina 
- YubiKey model and version: Yubikey 5C
- Bug description summary:

When ykman fails due to "No YubiKey Detected!" the stderr output includes usage text, even though the actual command is valid - it just can't be executed because there's no key to query.

I'm using ykman within a script to automatically pull TOTP codes, and ideally I'd like for it to report ONLY the error message on failure. Printing usage text for a hidden subcommand just ends up confusing things for the end user when they run my script. At the very least, it'd be nice to have an argument to suppress that part of the output.

Steps to reproduce
---

Run the following command with no yubikey connected:
```
$ ykman oath code -s "mycode"
```

Expected result
---

```
$ ykman oath code -s "mycode"
Error: No YubiKey detected!

```

Actual results and logs
---

```
$ ykman oath code -s "mycode"
Usage: ykman [OPTIONS] COMMAND [ARGS]...
Try "ykman -h" for help.

Error: No YubiKey detected!

```

Other info
---

It's worth noting that this behavior does NOT occur for other errors. Same command, this time with a Yubikey connected but no matching oath code configured:

```
$ ykman oath code -s "mycode"
Error: Multiple matches, please make the query more specific.

```

### PR DESCRIPTION
**Pull request by dagheyman**
Use a custom `cli_fail` function that doesn't print usage info.

Fixes #296

### CODE REVIEW

File: ykman/cli/__main__.py [hunk 1]
```diff
@@ -111,27 +111,27 @@ def _run_cmd_for_single(ctx, cmd, transports, reader=None):
...
-        ctx.fail('Multiple YubiKeys detected. Use --device SERIAL to specify '
+        cli_fail('Multiple YubiKeys detected. Use --device SERIAL to specify '
                  'which one to use.')
     descriptor = descriptors[0]
     if descriptor.mode.transports & transports:
         try:
             return descriptor.open_device(transports)
         except FailedOpeningDeviceException:
-            ctx.fail('Failed connecting to {} [{}]. Make sure the application have \
+            cli_fail('Failed connecting to {} [{}]. Make sure the application have \
```
Comment (emlun) [inline ykman/cli/__main__.py]: Unrelated, but typo:
```suggestion
            cli_fail('Failed connecting to {} [{}]. Make sure the application has \
```

File: ykman/cli/__main__.py [hunk 2]
```diff
@@ -177,12 +177,12 @@ def cli(ctx, device, log_level, log_file, reader):
         ykman.logging_setup.setup(log_level, log_file=log_file)
 
     if reader and device:
-        ctx.fail('--reader and --device options can\'t be combined.')
+        cli_fail('--reader and --device options can\'t be combined.')
```
Comment (emlun) [inline ykman/cli/__main__.py]: Maybe this should still print the usage info?

File: ykman/cli/__main__.py [hunk 3]
```diff
@@ -177,12 +177,12 @@ def cli(ctx, device, log_level, log_file, reader):
...
 
     subcmd = next(c for c in COMMANDS if c.name == ctx.invoked_subcommand)
     if subcmd == list_keys:
         if reader:
-            ctx.fail('--reader and list command can\'t be combined.')
+            cli_fail('--reader and list command can\'t be combined.')
```
Comment (emlun) [inline ykman/cli/__main__.py]: Maybe this should still print the usage info?

File: ykman/cli/config.py [hunk 4]
```diff
@@ -138,10 +139,10 @@ def set_lock_code(new_lock_code):
                 reboot=True)
         except Exception as e:
             logger.error('Setting the lock code failed', exc_info=e)
-            ctx.fail('Failed to set the lock code.')
+            cli_fail('Failed to set the lock code.')
 
     if generate and new_lock_code:
-        ctx.fail('Invalid options: --new-lock-code conflicts with --generate.')
+        cli_fail('Invalid options: --new-lock-code conflicts with --generate.')
```
Comment (emlun) [inline ykman/cli/config.py]: Maybe this should still print the usage info?

File: ykman/cli/config.py [hunk 5]
```diff
@@ -231,14 +232,14 @@ def ensure_not_all_disabled(ctx, usb_enabled):
...
-        ctx.fail('No configuration options chosen.')
+        cli_fail('No configuration options chosen.')
 
     enable = list(APPLICATION) if enable_all else enable
 
     _ensure_not_invalid_options(ctx, enable, disable)
 
     if touch_eject and no_touch_eject:
-        ctx.fail('Invalid options.')
+        cli_fail('Invalid options.')
```
Comment (emlun) [inline ykman/cli/config.py]: Maybe this should still print the usage info?

File: ykman/cli/config.py [hunk 6]
```diff
@@ -415,15 +416,15 @@ def _list_apps(ctx, enabled):
 
 def _ensure_not_invalid_options(ctx, enable, disable):
     if any(a in enable for a in disable):
-        ctx.fail('Invalid options.')
+        cli_fail('Invalid options.')
```
Comment (emlun) [inline ykman/cli/config.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 7]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
     controller = ctx.obj['controller']
 
     if public_id and serial_public_id:
-        ctx.fail('Invalid options: --public-id conflicts with '
+        cli_fail('Invalid options: --public-id conflicts with '
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 8]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
...
                  '--serial-public-id.')
 
     if private_id and generate_private_id:
-        ctx.fail('Invalid options: --private-id conflicts with '
+        cli_fail('Invalid options: --private-id conflicts with '
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 9]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
...
                  '--generate-public-id.')
 
     if upload and force:
-        ctx.fail('Invalid options: --upload conflicts with --force.')
+        cli_fail('Invalid options: --upload conflicts with --force.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 10]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
...
 
     if key and generate_key:
-        ctx.fail('Invalid options: --key conflicts with --generate-key.')
+        cli_fail('Invalid options: --key conflicts with --generate-key.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 11]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
...
 
     if not public_id:
         if serial_public_id:
             if dev.serial is None:
-                ctx.fail('Serial number not set, public ID must be provided')
+                cli_fail('Serial number not set, public ID must be provided')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 12]
```diff
@@ -262,29 +262,29 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
...
             public_id = modhex_encode(
                 b'\xff\x00' + struct.pack(b'>I', dev.serial))
             click.echo(
                 'Using YubiKey serial as public ID: {}'.format(public_id))
         elif force:
-            ctx.fail(
+            cli_fail(
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 13]
```diff
@@ -302,7 +302,7 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
                 'Using a randomly generated private ID: {}'.format(
                     b2a_hex(private_id).decode('ascii')))
         elif force:
-            ctx.fail(
+            cli_fail(
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 14]
```diff
@@ -316,7 +316,7 @@ def yubiotp(ctx, slot, public_id, private_id, key, no_enter, force,
                 'Using a randomly generated secret key: {}'.format(
                     b2a_hex(key).decode('ascii')))
         elif force:
-            ctx.fail('Secret key not given. Please remove the --force flag, or '
+            cli_fail('Secret key not given. Please remove the --force flag, or '
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 15]
```diff
@@ -425,14 +425,14 @@ def chalresp(ctx, slot, key, totp, touch, force, generate):
 
     if key:
         if generate:
-            ctx.fail('Invalid options: --generate conflicts with KEY argument.')
+            cli_fail('Invalid options: --generate conflicts with KEY argument.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 16]
```diff
@@ -382,9 +382,9 @@ def static(
     controller = ctx.obj['controller']
 
     if password and len(password) > 38:
-        ctx.fail('Password too long (maximum length is 38 characters).')
+        cli_fail('Password too long (maximum length is 38 characters).')
     if generate and not length:
-        ctx.fail('Provide a length for the generated password.')
+        cli_fail('Provide a length for the generated password.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 17]
```diff
@@ -425,14 +425,14 @@ def chalresp(ctx, slot, key, totp, touch, force, generate):
...
-            ctx.fail('Invalid options: --generate conflicts with KEY argument.')
+            cli_fail('Invalid options: --generate conflicts with KEY argument.')
         elif totp:
             key = parse_b32_key(key)
         else:
             key = parse_key(key)
     else:
         if force and not generate:
-            ctx.fail('No secret key given. Please remove the --force flag, '
+            cli_fail('No secret key given. Please remove the --force flag, '
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 18]
```diff
@@ -479,20 +479,20 @@ def calculate(ctx, slot, challenge, totp, digits):
     """
     controller = ctx.obj['controller']
     if not challenge and not totp:
-        ctx.fail('No challenge provided.')
+        cli_fail('No challenge provided.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/otp.py [hunk 19]
```diff
@@ -586,10 +586,10 @@ def settings(ctx, slot, new_access_code, delete_access_code, enter, pacing,
     controller = ctx.obj['controller']
 
     if (new_access_code is not None) and delete_access_code:
-        ctx.fail('--new-access-code conflicts with --delete-access-code.')
+        cli_fail('--new-access-code conflicts with --delete-access-code.')
```
Comment (emlun) [inline ykman/cli/otp.py]: Maybe this should still print the usage info?

File: ykman/cli/piv.py [hunk 20]
```diff
@@ -753,12 +753,12 @@ def change_management_key(
         no_prompt=force)
 
     if new_management_key and generate:
-        ctx.fail('Invalid options: --new-management-key conflicts with '
+        cli_fail('Invalid options: --new-management-key conflicts with '
```
Comment (emlun) [inline ykman/cli/piv.py]: Maybe this should still print the usage info?

File: ykman/cli/piv.py [hunk 21]
```diff
@@ -780,7 +780,7 @@ def change_management_key(
                         b2a_hex(new_management_key).decode('utf-8')))
 
         elif force:
-            ctx.fail('New management key not given. Please remove the --force '
+            cli_fail('New management key not given. Please remove the --force '
```
Comment (emlun) [inline ykman/cli/piv.py]: Maybe this should still print the usage info?

File: ykman/cli/piv.py [hunk 22]
```diff
@@ -944,26 +944,26 @@ def _ensure_authenticated(
...
-        ctx.fail('PIN verification failed.')
+        cli_fail('PIN verification failed.')
 
 
 def _authenticate(ctx, controller, management_key, mgm_key_prompt,
                   no_prompt=False):
     if not management_key:
         if no_prompt:
-            ctx.fail('Management key required.')
+            cli_fail('Management key required.')
```
Comment (emlun) [inline ykman/cli/piv.py]: Maybe this should still print the usage info?

**Comment by dainnilsson** [comment]
Re-implemented in cff37e91b9ea9a53a496b0b943cb8090013ec0af rather than dealing with conflicts.

**Pull request closed without merging**